### ISSUE TITLE
segfault at exit when using non null terminated strong for glfwSetWindowTitle()

### ISSUE THREAD

**Issue by JeremyBorys** [glfw/glfw#1933]
When using glfwSetWindowTitle() on X11 I stupidly passed a non null terminated string. This gets passed down to https://linux.die.net/man/3/xutf8setwmproperties which requires a null terminated string.

I believe this is a bug and we should check for null termination but its not clear to me where the best place to do it is or what limitations we should impose.

Even maybe just calling out out some documentation is good enough. with a C++ string example of something like:

```cpp
auto title = string("This is my title");
glfwSetWindowTitle(window, title.c_str());
```

**Comment by tombsar** [comment]
I'm afraid I disagree that this is a bug. Strings being NUL-terminated in C is fundamental to the language, and all sorts of things (not just glfw) will break if you pass around char pointers without due care and attention. It is not the job of the API documentation to cover usage from other languages such as C++.

I suppose it wouldn't hurt to add "null-terminated" to the API documentation wherever we use strings. Do you think that change would have helped you to have avoided the problem?

> we should check for null termination

I just can't think of any way to do this. From the point of view of a function receiving a char pointer, there's just not enough information to distinguish a terminated string from an unterminated one. Best case it looks like a longer string than intended with garbage bytes at the end (there will be a 0 byte *somewhere* in memory), worst case the code faults trying to read out of bounds.

Out of interest, what did the code look like that was creating a non-terminated string?

Aside: if you want type-safety, I suggest looking at C++ wrappers, rather than using the C API directly. https://github.com/Cvelth/vkfw for example accepts a `std::string_view` rather than a raw pointer (if supported by the compiler), which probably would have avoided this issue.

**Comment by JeremyBorys** [comment]
Yeah. This is my mistake and I apologize for wasting your time.

It came about of a poor bisect while looking for an introduced segfault. The bisect identified a change made (by me) wrt to setting a window title. Looking at it further I found that it was something entirely different.

> Out of interest, what did the code look like that was creating a non-terminated string?

The code that I introduced looked similar to:

```cpp
setTitle(GLFWwindow* window, const std::string& name)
{
      auto title = "Some Title - " + name;

     /* I need to review the C++ docs / tests to see if data() returns a null terminated string.... I don't think there is any difference between .c_str() and .data()... */ 
      glfwSetWindowTitle(window, title.data());
}
```

> I suppose it wouldn't hurt to add "null-terminated" to the API documentation wherever we use strings. Do you think that change would have helped you to have avoided the problem?

I think this was the only thing that I could suggest that makes any sense. I don't think it will stop me personally from being dumb. But while debugging and tracing the docs after something weird happens, it will stop me from wasting your time because you already thought about it. I normally expect a small section about string expectations like UTF-8 encoded, always null terminated etc.

> Aside: if you want type-safety, I suggest looking at C++ wrappers, rather than using the C API directly. https://github.com/Cvelth/vkfw for example accepts a std::string_view rather than a raw pointer (if supported by the compiler), which probably would have avoided this issue.

Yeah, this is what I normally use for personal projects. The company I work for has already brought in the C-API instead and I personally don't care which tool we use C or C++.

**Comment by ReeceSX** [comment]
No, that isn't a C++ example of the bug; despite `std::string` being notoriously lawful, `.c_str()` guarantees null termination. Your initial example contradicts the one where you gloss over the comment `I don't think there is any difference between .c_str() and .data()`. They are not identical; `.data` does not guarantee null termination.

**Comment by JeremyBorys** [comment]
> Your initial example contradicts the one where you gloss over the comment

Sorry if I wasn't clear. I inserted that comment as I was moving it into the github comment.

> They are not identical; .data does not guarantee null termination.

Yeah, I was pretty sure that was the case and part of the motivation for this ticket.

The title "segfault at exit when using non null terminated strong for glfwSetWindowTitle()" is wrong after I did more digging and I need to rename it.

Since we have a C-API it is fair to say that this isn't a bug and a won't fix. It is fair to assume people understand that this is a c-str and strings should be null terminated. 

It is possible that that someone could be as dumb as I and create a similar issue again in the future. When this happens its likely we would have this discussion again. The alternative is a one line comment pointing to some documentation saying look here dummy.

Either solution is fine by me.
